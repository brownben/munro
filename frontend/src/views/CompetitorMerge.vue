<!--
  Merge Competitors

  Allows to merge two comopetitors together (when a duplicate)
-->

<template>
  <div>
    <vue-headful
      title="Munro - Merge Competitors"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <h1>Merge Competitors</h1>
    <form @submit.prevent="merge()">
      <label>League:</label>
      <dropdown-input
        v-model="league"
        :list="leagues.map(league => league.name)"
        :include-blank="true"
      />
      <label>Course:</label>
      <dropdown-input v-model="course" :list="coursesInLeague" :include-blank="true" />
      <label>Competitor to Keep:</label>
      <dropdown-input
        v-model="competitorKeep"
        :list="competitorsForLeague.map(competitor => competitor.id+' - '+competitor.name+' - '+competitor.ageClass+' - '+competitor.club)"
        :include-blank="true"
      />
      <label>Competitor to be Merged:</label>
      <dropdown-input
        v-model="competitorMerge"
        :list="competitorsForLeague.map(competitor => competitor.id+' - '+competitor.name+' - '+competitor.ageClass+' - '+competitor.club)"
        :include-blank="true"
      />
      <button>Merge Competitors</button>
    </form>
  </div>
</template>

<script>
import axios from 'axios'
import DropdownInput from '@/components/DropdownInput'

export default {
  components: {
    DropdownInput: DropdownInput,
  },

  data: () => ({
    competitors: [],
    leagues: [],
    league: '',
    competitorMerge: '',
    competitorKeep: '',
    course: '',
  }),

  computed: {
    competitorsForLeague: function () {
      const filtered = this.competitors.filter(competitor => competitor.league === this.league && competitor.course === this.course)
      return filtered.sort((a, b) => a.name > b.name)
    },

    coursesInLeague: function () {
      const selectedLeague = this.leagues.filter(league => league.name === this.league)
      if (selectedLeague.length > 0) return selectedLeague[0].courses
      else return []
    },
  },

  mounted: async function () {
    await this.getCompetitors()
    await this.getLeagues()
  },

  methods: {
    getCompetitors: function () {
      return axios.get('/api/competitors')
        .then(response => { this.competitors = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching Competitors'))
    },

    getLeagues: function () {
      return axios.get('/api/leagues')
        .then(response => { this.leagues = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching League Details'))
    },

    validateForm: function () {
      return (
        this.competitorMerge !== this.competitorKeep &&
        this.competitorMerge !== '' &&
        this.competitorKeep !== '')
    },

    merge: function () {
      if (this.validateForm()) {
        return axios.post('/api/competitors/merge', {
          competitorKeep: this.competitorKeep.split(' -')[0],
          competitorMerge: this.competitorMerge.split(' -')[0],
        })
          .then(response => this.returnToCompetitorsPage(response))
          .catch(error => this.$messages.addMessage(error.response.data.message))
      }
      else this.$messages.addMessage('The Competitors Selected are the Same')
    },

    returnToCompetitorsPage: function (response) {
      // Go to league page after successful update/ creation
      this.$messages.addMessage(response.data.message)
      this.$router.push('/competitors')
    },

    competitorTransformForSelect: function (competitor) {
      if (competitor.club && competitor.ageClass) return competitor.name + ' (' + competitor.ageClass + ', ' + competitor.club + ') [' + competitor.id + ']'
      else if (competitor.club) return competitor.name + ' (' + competitor.club + ') [' + competitor.id + ']'
      else if (competitor.ageClass) return competitor.name + ' (' + competitor.ageClass + ') [' + competitor.id + ']'
      else return competitor.name + ' [' + competitor.id + ']'
    },
  },
}
</script>
<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'

h1
  margin-bottom: 0.5rem
</style>
