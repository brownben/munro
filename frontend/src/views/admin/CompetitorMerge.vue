<template>
  <Layout title="Merge Competitors">
    <Meta
      title="Munro - Merge Competitors"
      description=""
      :block-robots="true"
    />
    <form class="col-span-2" @submit.prevent="mergeCompetitors">
      <InputDropdown
        v-model="choices.league"
        :list="leagues.map((league) => league.name)"
        :include-blank="true"
        label="League:"
        url-parameter="league"
      />
      <InputDropdown
        v-model="choices.course"
        :list="courses"
        :include-blank="true"
        label="Course"
        class="mt-4"
        url-parameter="course"
      />
      <InputDropdown
        v-model="choices.competitorKeep"
        :listWithDifferentValue="
          competitorsInLeague.map((competitor) => ({
            value: competitor.id.toString(),
            text: competitorToText(competitor),
          }))
        "
        :include-blank="true"
        label="Competitor to Keep:"
        class="mt-4"
        url-parameter="competitorToKeep"
      />
      <InputDropdown
        v-model="choices.competitorMerge"
        :listWithDifferentValue="
          competitorsInLeague.map((competitor) => ({
            value: competitor.id.toString(),
            text: competitorToText(competitor),
          }))
        "
        :include-blank="true"
        label="Competitor to be Merged:"
        class="mt-4"
      />
      <button class="mt-8 button-lg">Merge Competitors</button>
    </form>
  </Layout>
</template>

<script lang="ts" setup>
import { ref, onMounted, computed } from 'vue'
import { useRouter } from 'vue-router'
import { useStore } from 'vuex'

import Layout from '../../components/Layout.vue'
import InputDropdown from '../../components/InputDropdown.vue'

import { useLeagues } from '../../api/leagues'
import {
  useCompetitors,
  mergeCompetitors as apiMergeCompetitors,
} from '../../api/competitors'

const store = useStore()
const router = useRouter()

const [leagues] = useLeagues()
const [competitors] = useCompetitors()

const choices = ref({
  league: '',
  competitorMerge: '',
  competitorKeep: '',
  course: '',
})

const courses = computed(
  () =>
    leagues.value?.find((league) => league.name === choices.value.league)
      ?.courses ?? []
)
const competitorsInLeague = computed(() =>
  competitors.value
    ?.filter(
      (competitor) =>
        competitor.league === choices.value.league &&
        competitor.course === choices.value.course
    )
    ?.sort((a, b) => (a.name > b.name ? 1 : -1))
)

const competitorToText = (competitor: Competitor) => {
  if (competitor.club && competitor.ageClass)
    return `${competitor.name} (${competitor.ageClass}, ${competitor.club}) [${competitor.id}]`
  else if (competitor.club)
    return `${competitor.name} (${competitor.club}) [${competitor.id}]`
  else if (competitor.ageClass)
    return `${competitor.name} (${competitor.ageClass}) [${competitor.id}]`
  else return `${competitor.name} [${competitor.id}]`
}

const validateForm = () => {
  if (
    choices.value.competitorMerge !== choices.value.competitorKeep &&
    choices.value.competitorMerge !== '' &&
    choices.value.competitorKeep !== ''
  )
    return true
  else {
    store.dispatch(
      'createMessage',
      'Please Ensure Competitors Are Not The Same'
    )
    return false
  }
}

const mergeCompetitors = () => {
  if (validateForm())
    apiMergeCompetitors({
      competitorMerge: Number(choices.value.competitorMerge),
      competitorKeep: Number(choices.value.competitorKeep),
    })
      .then(() => router.push(`/competitors/${choices.value.competitorKeep}`))
      .catch(() => false)
}
</script>
