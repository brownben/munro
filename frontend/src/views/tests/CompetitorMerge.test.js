import { mount } from '@vue/test-utils'
import CompetitorMerge from '@/views/CompetitorMerge'

test('Is a Vue Instance', () => {
  const wrapper = mount(CompetitorMerge)
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = mount(CompetitorMerge)
  expect(wrapper.element).toMatchSnapshot()
})

