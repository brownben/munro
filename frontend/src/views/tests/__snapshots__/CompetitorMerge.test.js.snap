// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div
  class="view"
>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Merge Competitors"
  />
   
  <h1
    class="text-main text-3xl font-normal font-heading mb-2"
  >
    Merge Competitors
  </h1>
   
  <form>
    <dropdown-input-stub
      include-blank="true"
      label="League:"
      list=""
      value=""
    />
     
    <dropdown-input-stub
      include-blank="true"
      label="Course:"
      list=""
      value=""
    />
     
    <dropdown-input-stub
      include-blank="true"
      label="Competitor to Keep:"
      list=""
      value=""
    />
     
    <dropdown-input-stub
      include-blank="true"
      label="Competitor to be Merged:"
      list=""
      value=""
    />
     
    <button
      class="button-lg"
    >
      Merge Competitors
    </button>
  </form>
</div>
`;
